go/pkg/cas/upload.go

```diff
@@ -39,6 +59,358 @@ type DigestStat struct {
+	isLarge := info.Size() >= u.LargeFileThreshold
+
+	// Lock the mutex before acquiring a semaphore to avoid hogging the latter.
+	if isLarge {
+		// Read only one large file at a time.
+		u.muLargeFile.Lock()
+		defer u.muLargeFile.Unlock()
+	}
+
+	if err := u.fsSem.Acquire(ctx, 1); err != nil {
+		return nil, err
+	}
+	defer u.fsSem.Release(1)
+
+	f, err := os.Open(absPath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	ret := &repb.FileNode{
+		Name:         info.Name(),
+		IsExecutable: (info.Mode() & 0100) != 0,
+	}
+
+	if info.Size() <= u.SmallFileThreshold {
+		// This file is small enough to buffer it entirely.
+		contents, err := ioutil.ReadAll(f)
+		if err != nil {
+			return nil, err
+		}
+		item := uploadItemFromBlob(absPath, contents)
+		ret.Digest = item.Digest
+		return ret, u.scheduleCheck(ctx, item)
+	}
+
+	// It is a medium or large file.
+
+	// Compute the hash.
+	// TODO(nodir): reuse the buffer.
+	buf := make([]byte, u.FileIOSize)
+	dig, err := digest.NewFromReaderWithBuffer(f, buf)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to compute hash")
+	}
+	ret.Digest = dig.ToProto()
+
+	item := &uploadItem{
+		Title:  absPath,
+		Digest: ret.Digest,
+	}
+
+	if isLarge {
+		// Large files are special: locality is important - we want to re-read the
+		// file ASAP.
+		// Also we are not going to use BatchUploads anyway, so we can take
+		// advantage of ByteStream's built-in presence check.
+		// https://github.com/bazelbuild/remote-apis/blob/0cd22f7b466ced15d7803e8845d08d3e8d2c51bc/build/bazel/remote/execution/v2/remote_execution.proto#L250-L254
+
+		item.Open = func() (readSeekCloser, error) {
+			_, err := f.Seek(0, io.SeekStart)
+			return f, err
+		}
+		panic("not implemented")
+		// TODO(nodir): implement.
+	}
+
+	// Schedule a check and close the file (in defer).
+	// item.Open will reopen the file.
+
+	item.Open = func() (readSeekCloser, error) {
+		return os.Open(absPath)
+	}
+	return ret, u.scheduleCheck(ctx, item)
+}
+
+// visitDir reads a directory and its descendants, subject to the
+// predicate. The function blocks until each descendant is visited, but the
+// visitation happens concurrently, using u.eg.
+func (u *uploader) visitDir(ctx context.Context, absPath string) (*repb.DirectoryNode, error) {
+	var mu sync.Mutex
+	dir := &repb.Directory{}
+	var subErr error
+	var wg sync.WaitGroup
+
+	// This sub-function exist to avoid holding the semaphore while waiting for
+	// children.
+	err := func() error {
+		if err := u.fsSem.Acquire(ctx, 1); err != nil {
+			return err
+		}
+		defer u.fsSem.Release(1)
+
+		f, err := os.Open(absPath)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		// Check the context, since file IO functions don't.
+		for ctx.Err() == nil {
+			infos, err := f.Readdir(128)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return err
+			}
+
+			for _, info := range infos {
+				info := info
+				absChild := filepath.Join(absPath, info.Name())
```

Comment:
thinking: filepath functions _always_ call `Clean`, even when we know it isn't necessary (absPath should have been cleaned on `startProcessing`; AFAICT `info.Name()` should be as good as clean?). 

At the same time using `strings.Join(..., os.Separator)` feels awkward? But that's what filepath does: https://golang.org/src/path/filepath/path_unix.go. wdyt of using it directly?

Reply:
Added a new func. PTAL. `strings.Join` seems overkill here.